### ISSUE TITLE
Theorems which don't use some of their hypotheses

### ISSUE THREAD

**Issue by savask** [metamath/set.mm#2648]
A recent pull request #2635 mentioned a theorem `rspcda` which didn't use one of its essential hypotheses. I ran a program to find other examples of such theorems in set.mm, but since there are hundreds of occurrences, I narrowed down the search with the following rules:

* A theorem must be in the main part of set.mm,
* It is not discouraged,
* A theorem and all its essential hypotheses are enclosed in a block, and this block contains one theorem only.

There are 53 such theorems, here's the list grouped by contributors:

**AV**: ~~[chmaidscmat](http://us.metamath.org/mpeuni/chmaidscmat.html), [clwwlkvbij](http://us.metamath.org/mpeuni/clwwlkvbij.html), [cramerimplem1](http://us.metamath.org/mpeuni/cramerimplem1.html), [distspace](http://us.metamath.org/mpeuni/distspace.html), [edgnbusgreu](http://us.metamath.org/mpeuni/edgnbusgreu.html), [fnmpt2ovd](http://us.metamath.org/mpeuni/fnmpt2ovd.html), [gsummptnn0fz](http://us.metamath.org/mpeuni/gsummptnn0fz.html), [inveq](http://us.metamath.org/mpeuni/inveq.html), [matmulcell](http://us.metamath.org/mpeuni/matmulcell.html), [mdet0](http://us.metamath.org/mpeuni/mdet0.html), [minmar1marrep](http://us.metamath.org/mpeuni/minmar1marrep.html), [ressval3d](http://us.metamath.org/mpeuni/ressval3d.html), [uhgrvtxedgiedgb](http://us.metamath.org/mpeuni/uhgrvtxedgiedgb.html), [uhgrwkspth](http://us.metamath.org/mpeuni/uhgrwkspth.html), [ushgredgedgloop](http://us.metamath.org/mpeuni/ushgredgedgloop.html), [zrhcofipsgn](http://us.metamath.org/mpeuni/zrhcofipsgn.html)~~
**JH**: ~~[2ndcctbss](http://us.metamath.org/mpeuni/2ndcctbss.html)~~
**MC**: [axdc3lem3](http://us.metamath.org/mpeuni/axdc3lem3.html), [bpos1lem](http://us.metamath.org/mpeuni/bpos1lem.html), [coeeulem](http://us.metamath.org/mpeuni/coeeulem.html), [fucidcl](http://us.metamath.org/mpeuni/fucidcl.html), [gsumzsubmcl](http://us.metamath.org/mpeuni/gsumzsubmcl.html), [haushmphlem](http://us.metamath.org/mpeuni/haushmphlem.html), [infxpenc](http://us.metamath.org/mpeuni/infxpenc.html), ~~[itgulm2](http://us.metamath.org/mpeuni/itgulm2.html), [pilem2](http://us.metamath.org/mpeuni/pilem2.html)~~, [seqcaopr3](http://us.metamath.org/mpeuni/seqcaopr3.html), [tsmslem1](http://us.metamath.org/mpeuni/tsmslem1.html), [txcnpi](http://us.metamath.org/mpeuni/txcnpi.html), [vdwlem1](http://us.metamath.org/mpeuni/vdwlem1.html), [vdwlem2](http://us.metamath.org/mpeuni/vdwlem2.html), [vdwlem8](http://us.metamath.org/mpeuni/vdwlem8.html), [xrge0gsumle](http://us.metamath.org/mpeuni/xrge0gsumle.html), ~~[xrhmeo](http://us.metamath.org/mpeuni/xrhmeo.html)~~
**NM**: ~~[axpweq](http://us.metamath.org/mpeuni/axpweq.html), [brecop2](http://us.metamath.org/mpeuni/brecop2.html), [euind](http://us.metamath.org/mpeuni/euind.html), [karden](http://us.metamath.org/mpeuni/karden.html), [lspfixed](http://us.metamath.org/mpeuni/lspfixed.html), [lspsncv0](http://us.metamath.org/mpeuni/lspsncv0.html), [lssneln0](http://us.metamath.org/mpeuni/lssneln0.html), [lssssr](http://us.metamath.org/mpeuni/lssssr.html), [opeqsn](http://us.metamath.org/mpeuni/opeqsn.html), [uzind4i](http://us.metamath.org/mpeuni/uzind4i.html)~~
**PC**: ~~[alginv](http://us.metamath.org/mpeuni/alginv.html)~~
**SF**: ~~[wfrlem4](http://us.metamath.org/mpeuni/wfrlem4.html)~~
**SO**: [deg1sublt](http://us.metamath.org/mpeuni/deg1sublt.html), [gsumvsmul](http://us.metamath.org/mpeuni/gsumvsmul.html), [psgnunilem1](http://us.metamath.org/mpeuni/psgnunilem1.html), [zrhpsgninv](http://us.metamath.org/mpeuni/zrhpsgninv.html)
**TA**: ~~[gsummptun](http://us.metamath.org/mpeuni/gsummptun.html)~~, [mpteq12df](http://us.metamath.org/mpeuni/mpteq12df.html), [tglnpt2](http://us.metamath.org/mpeuni/tglnpt2.html)

Most of the unused hypotheses are of the form "some variable lies in some class", for example `A e. _V`, but there are more interesting examples like [pilem2](http://us.metamath.org/mpeuni/pilem2.html), which doesn't use the hypothesis `_pi < A`.

Since each theorem in the list lies in its own block, I think it should be possible to simply delete unused hypotheses and fix all affected theorems automatically with metamath.exe. Needless to say it will reduce set.mm size both directly by removing unused lines and indirectly by making compressed proofs smaller. What is your opinion? (@jkingdon @benjub @tirix @avekens)

**Comment by benjub** [comment]
Thanks @savask, that's very nice !
To update automatically the proofs, a way would be to save the proofs of the theorems in question and the theorems using them in the normal format, then delete the extraneous hypotheses, and then save all proofs in the compressed format. Something like:
```
MM> read set.mm
MM> open log 1.tmp
MM> show usage chmaidscmat
MM> close log
MM> submit save_normal.cmd
```
then delete the hypotheses, and then save everything back to compressed format (as in https://github.com/metamath/set.mm/blob/develop/CONTRIBUTING.md). Actually, if you have enough memory, it might be possible to save everything in normal format. If not, then do as above and you can write save_normal.cmd on the model of scripts/min.cmd. Further automation is possible...

This may make that some new proofs do not use some hypotheses, so you could rerun your program to see if that happens.  It would be nice to run it on all theorems (i.e., mathboxes, discouraged, and shared within blocks), and also on other databases (mainly iset.mm).

You might share your program, maybe by pushing it in `scripts/`.  It's nice to have a growing suite of "sanitizing" programs that can be run periodically (e.g., yearly) like this one or those developed by @jamesjer looking for extraneous braces and DVs (these might be run since your simplifications may create some of these (edit: an example is http://us.metamath.org/mpeuni/2ndcctbss.html where `X` simply disappears)).  Another nice program would be a more aggressive elimination of extraneous DVs.

**Comment by avekens** [comment]
Very interesting result! I will have a look at my 16 concerned theorems and hopefully can clean them up with my next PR(s).

**Comment by jkingdon** [comment]
> more interesting examples like [pilem2](http://us.metamath.org/mpeuni/pilem2.html), which doesn't use the hypothesis `_pi < A`.

There's something odd going on with this one. Unless I am misreading this theorem, the hypothesis isn't even true (the hypotheses taken together say that A is a real number between 2 and 4 whose sine is zero and which is greater than pi). I think this is saying this set of lemmas could be simplified further, but getting rid of the unused hypothesis could be a good next step so go for it.

**Comment by benjub** [comment]
> There's something odd going on with this one. Unless I am misreading this theorem, the hypothesis isn't even true (the hypotheses taken together say that A is a real number between 2 and 4 whose sine is zero and which is greater than pi).

Good catch ! (It sounds weird to say "the hypothesis isn't true"... you mean that the hypotheses are incompatible.) One could have proved `${ ... $p |- F. $= ? $. $}` from them ! (**Edit**: no, as @digama0 wrote below, the common antecedent of the deduction, phi, has to be false, and then everything is vacuously true.)

> I think this is saying this set of lemmas could be simplified further

I don't see that... (which does not mean these lemmas cannot be simplified further, only that the first part of your comment does not obviously imply it).

**Comment by digama0** [comment]
If the hypotheses to pilem2 are contradictory, then I don't see how it could be a lemma... is it unused?

EDIT: Oh wait I remember now, the hypotheses are supposed to be contradictory. It's part of a proof by contradiction that pi is indeed the unique solution to those constraints.

I think the reason why the hypothesis `pi < A` is there is simply because it was available in pilem3 so it was passed along. The proof is not much changed by its absence.

**Comment by benjub** [comment]
When pilem2 is used (in pilem3), the antecedent is substituted with
```
 ( ( ( x e. ( 2 (,) 4 ) /\ ( sin ` x ) = 0 ) /\ _pi < x ) /\ y e. ( RR+ i^i ( `' sin " { 0 } ) ) )
```
Maybe one can remove the conjunct `_pi < x ` ?

**Comment by jkingdon** [comment]
> It's part of a proof by contradiction that pi is indeed the unique solution to those constraints.

Agreed. Sorry if my comment was confusing because I didn't state this directly.

As for whether removing this hypothesis leads to further simplifications, I agree that it isn't self evident as it depends on where else pi < A is used. So I was a bit jumping to conclusions on that part.

**Comment by jkingdon** [comment]
> What is your opinion?

Although I could speculate on how these hypotheses got there, I can't think of any reason to keep them.

**Comment by digama0** [comment]
Looking at the proof (in pilem3), I think it is true that the pi < A part can be completely removed (it appears in step 92 of http://us.metamath.org/mpeuni/pilem3.html that applies pm2.18d); you can just prove x <= pi directly instead.

In any case, I think it is unquestionable that removing the extra hypotheses from the marked theorems without changing downstream theorems will lead to shortening, and it may or may not allow the downstream proofs to be simplified as well (like pilem3).

**Comment by savask** [comment]
@benjub 
>To update automatically the proofs, a way would be to save the proofs of the theorems in question and the theorems using them in the normal format, then delete the extraneous hypotheses, and then save all proofs in the compressed format ...

While this will work for the theorem itself, I don't think it will work for the affected theorems (the ones from `show usage`). This is because the normal format proof will still contain the sequence of labels putting a useless hypothesis on the stack.

The only way to update the proofs automatically I could come up with is as follows:
1) Rename `theorem` into `theoremALT` including all labels used in proofs
2) Copy `theoremALT` into `theorem`, remove the useless hypothesis from `theorem` and decompress/compress the proof as you suggested
3) Minimize with `theorem`, removing all uses of `theoremALT`
4) Delete `theoremALT`

> It would be nice to run it on all theorems (i.e., mathboxes, discouraged, and shared within blocks), and also on other databases (mainly iset.mm).

If I run it on all theorems, I get 1182 hits which strongly hints at the need for further automation :-) While it should be relatively easy to automate the process I described above for theorems which lie in their personal blocks, it doesn't seem so easy in other cases. For example, theorem [lfgredgge2](http://us.metamath.org/mpeuni/lfgredgge2.html) doesn't use its first two hypotheses, but the next theorem in the same block does, see [lfgrnloop](http://us.metamath.org/mpeuni/lfgrnloop.html). This can be fixed by putting unused hypotheses after `lfgredgge2`, so our imaginary automated tool should be able to sort hypotheses within one block according on their usage.

>You might share your program, maybe by pushing it in scripts/.

It's a dirty hack built on top of my own metamath verifier (in Haskell), so it does not belong to `scripts/`. It would be great if metamath.exe or metamath-knife could do such sanitation jobs instead.

**Comment by jamesjer** [comment]
The linked PR above (https://github.com/metamath/metamath-exe/pull/92) does unused hypothesis and dv checking.  It's for metamath.exe, which I know is in maintenance mode, but C is still my strongest language.  I'm working my way through the Rust tutorial and hope it won't be much longer before I can contribute meaningfully to metamath-knife.

**Comment by benjub** [comment]
> The only way to update the proofs automatically I could come up with is as follows:

Ah yes, you're right. It is actually the way you indicate that has been used a few times already.  I don't know how easy it will be to automate.  Also, if the hypotheses are xxx.1... xxx..4 and xxx.2 is unued, we might want to rename xxx.3 and xxx.4 (but if too difficult, it's not a big deal).

Indeed, the theorems sharing $e-hypotheses with others might have to be done manually. We could concentrate on the earliest ones and/or the ones we deem more important. I'd be interested if you scanned my mathbox !

**Comment by savask** [comment]
@jamesjer 
This is great! I think metamath.exe will stay our most powerful tool for database manipulation for quite some time, so your effort surely didn't go to waste.

@benjub 
Unfortunately my program doesn't store assertions in the order they are present in set.mm, so I cannot easily run my tool on a specific theorem range. Here's a full list of "bad theorems": https://gist.github.com/savask/7e1ecaee57fcadc749c483c2082f0b29 Perhaps you can somehow extract theorem labels used in your mathbox, and check if any of them appear in this list (I couldn't do it myself, but I already see several theorems with bj- prefix in the list).

**Comment by benjub** [comment]
> Unfortunately my program doesn't store assertions in the order they are present in set.mm, so I cannot easily run my tool on a specific theorem range. Here's a full list of "bad theorems": https://gist.github.com/savask/7e1ecaee57fcadc749c483c2082f0b29 Perhaps you can somehow extract theorem labels used in your mathbox, and check if any of them appear in this list (I couldn't do it myself, but I already see several theorems with bj- prefix in the list).

Thanks ! I will fix the two theorems in my mathbox with extraneous hypotheses (and see how they crept in and if this simplifies other theorems). The other two are actually voluntary... I'll see what means I can employ to make them artifically appear in the proof.

Does your list also include cases with shared hypotheses and discouraged proofs/theorems ?

**Comment by savask** [comment]
>Does your list also include cases with shared hypotheses and discouraged proofs/theorems ?

Yes, everything.

**Comment by benjub** [comment]
> Yes, everything.

Perfect ! I looked at 15 to 20 of them, and it looks sometimes like an indication that the proof doesn't go exactly as the contributor intended.  This is interesting, and a reason to investigate further. So maybe wait a bit before a global automatic treatment, and I encourage contributors to see if the they recognize in the list some significant labels that they want to have a closer look at.

I'll try to order the list by order of occurrence in set.mm (probably a handful of lines of script, e.g. append " $p" at the end of each line of your file and order by first occurrence in set.mm), so as to concentrate on the earliest ones. If you have a handy way of having contributors on each line too, this could help people concentrate on "their" theorems (without exclusion of the others').

**Comment by savask** [comment]
> So maybe wait a bit before a global automatic treatment, and I encourage contributors to see if the they recognize in the list some significant labels that they want to have a closer look at.

I agree, there are also theorems which contain useless hypotheses by design, like `a1ii`, which shouldn't be corrected.

>If you have a handy way of having contributors on each line too, this could help people concentrate on "their" theorems (without exclusion of the others').

I do not, unfortunately. For the original list from the top post I had to do it by hand.

**Comment by benjub** [comment]
Another instance where probably one should not remove the hypothesis is nfof/nfofr, by @digama0, which does not use its hypothesis, but the intention was probably to use it and avoid the DV condition which crept up (this is probably a result of mmj2 which automatically fills in proofs with theorems which may require DV conditions, and then adds the DV conditions on its own).

I have sorted the theorems by earliest occurrence, and there are some pretty early examples, including two where two $e-hypotheses are unused. I don't know how easy it is to upload to gist, but the sorted list is the ouput of this script:
```
#!/usr/bin/env python3

"""
Sort a list of $p-labels by earliest occurrence
"""


with open('waste-thms.txt') as f:
    labels = {line.rstrip('\n') for line in f}
n = i = 0
with open('set.mm') as f:
    for line in f:
        toks = line.split()
        if len(toks) > 1 and toks[1] == '$p':
            n += 1
            label = toks[0]
            if label in labels:
                i += 1
                print(label)
                labels.remove(label)
print('Total number of $p-stmts:', n)
print('Number of wasteful $p-stmts:', i)
```

**Comment by digama0** [comment]
nfof looks way too suspicious to be a product of mmj2. I surely would have noticed if the entire proof was missing! I think it is more likely that a botched minimize run shortened the proof (I think there is an option to MINIMIZE to allow new DV conditions).

**Comment by benjub** [comment]
pilem2 and its unique usage pilem3: done (following indications by my and @digama0's posts above).

**Comment by jkingdon** [comment]
> nfof looks way too suspicious to be a product of mmj2. I surely would have noticed if the entire proof was missing! I think it is more likely that a botched minimize run shortened the proof (I think there is an option to MINIMIZE to allow new DV conditions).

Since no one got to this first, I submitted a pull request to fix this at https://github.com/metamath/set.mm/pull/2661

I didn't dig into the git history looking for clues about how this happened, but I guess my default assumption is that whatever went wrong was with a tool or process which is now fixed (given that I think we've made improvements in this sort of area over time).

**Comment by benjub** [comment]
> I didn't dig into the git history looking for clues about how this happened, but I guess my default assumption is that whatever went wrong was with a tool or process which is now fixed (given that I think we've made improvements in this sort of area over time).

It looks like a DV condition from an earlier theorem "leaked" to these because a {...}-bloc was not closed. Long blocs allow to avoid rewriting some DV conditions and hypotheses, but they come with a risk.

**Comment by avekens** [comment]
After I cleaned up my theorems now (see PR #2672), I will have a look at Norm's (remaining) theorems next:
NM: [brecop2](http://us.metamath.org/mpeuni/brecop2.html), [karden](http://us.metamath.org/mpeuni/karden.html), [lspfixed](http://us.metamath.org/mpeuni/lspfixed.html), [lspsncv0](http://us.metamath.org/mpeuni/lspsncv0.html), [lssneln0](http://us.metamath.org/mpeuni/lssneln0.html), [lssssr](http://us.metamath.org/mpeuni/lssssr.html), [opeqsn](http://us.metamath.org/mpeuni/opeqsn.html), [uzind4i](http://us.metamath.org/mpeuni/uzind4i.html)

**Comment by benjub** [comment]
@avekens : I've had a quick look and some of them seem to have ramifications (in the sense that some superfluous hypotheses "propagate" to theorems using them). So don't hesitate to separate them into several PRs.

**Comment by avekens** [comment]
> After I cleaned up my theorems now (see PR #2672), I will have a look at Norm's (remaining) theorems next:
> NM: [brecop2](http://us.metamath.org/mpeuni/brecop2.html), [karden](http://us.metamath.org/mpeuni/karden.html), [lspfixed](http://us.metamath.org/mpeuni/lspfixed.html), [lspsncv0](http://us.metamath.org/mpeuni/lspsncv0.html), [lssneln0](http://us.metamath.org/mpeuni/lssneln0.html), [lssssr](http://us.metamath.org/mpeuni/lssssr.html), [opeqsn](http://us.metamath.org/mpeuni/opeqsn.html), [uzind4i](http://us.metamath.org/mpeuni/uzind4i.html)

I had a closer look at these theorems, and I think the cleanup of the following theorems are straightforward:

* brecop2,
** `brecop2.1 $e |- .~ e. _V $. `superfluous
** used in ltsrpr (only proof affected)
* karden,
**  ` karden.2 $e |- B e. _V $. `superfluous
** not used
* lspfixed,
** ` lspfixed.x $e |- ( ph -> X e. V ) $. ` superfluous
** used in lsatfixedN  (only proof affected)
* lspsncv0,
** ` lspsncv0.e $e |- ( ph -> X =/= .0. ) $. `  superfluous
** used in lsatcv0  (only proof affected)
* lssssr,
** ` lssssr.v $e |- V = ( Base ` W ) $. ` superfluous
** used in dihjat1lem (only proof affected)
* uzind4i: 
** ` uzind4i.1 $e |- M e. ZZ $. ` superfluous
** used in om2uzrani, chtub, bclbnd (only proofs affected)

I plan to clean up these theorems in one of my next PRs. The following theorems need further investigation:

* opeqsn: This theorem (and the theorems which are based on this theorem) depends on the definition of an ordered pair (the hint "(Avoid depending on this detail.)" should be added to the comment (as in ~opeqpr), shouldn't it?).
* lssneln0: this theorem is used in 20 proofs

**Comment by benjub** [comment]
Thanks @avekens. Please do a specific PR for these.

**Comment by benjub** [comment]
> * opeqsn: This theorem (and the theorems which are based on this theorem) depends on the definition of an ordered pair (the hint "(Avoid depending on this detail.)" should be added to the comment (as in ~opeqpr), shouldn't it?).

It should (and skimming through some usages of opeqsn, some (most?) of them should too).

**Comment by david-a-wheeler** [comment]
@savask : great work identifying these cases!

As noted earlier:

> ... there are also theorems which contain useless hypotheses by design, like a1ii, which shouldn't be corrected.

I just posted to the mailing list that that we should use these comment markings for assertions ($a and $p) where hypotheses are not used on purpose:

* "(Intentionally unused hypothesis)" if exactly one hypothesis is unused
* "(Intentionally unused hypotheses)" if more than one hypotheses are unused

This is intentionally similar to other markings like (New usage is discouraged.)

**Comment by avekens** [comment]
~ wfrlem4 was also revised with PR #2694

**Comment by icecream17** [comment]
I am removing hypotheses often enough that I should probably work on making an automatic version of this.

Here's a preliminary specification
```
remove-hyp <database> <theorem> <hypotheses>

0. Check if there are theorems using <theorem> within the same block. If so, return an error saying to move <theorem> outside of the block manually before trying again.
1. Copy <theorem> into the next available global scope space, but deleting <hypotheses>, taking metamath-exe's normal proof for the old theorem, pasting it into this new theorem, and compressing.
2. Rename the original <theorem> into some arbitrary string of the same length (notated here as <abcdefg>); replacing it in the theorem itself and all proofs, but not comments
3. Call metamath.exe to minimize all uses of <abcdefg> into <theorem>
4. Delete <abcdefg> (no OLD version since it's automatic)
```

Unfortunately, `./scripts/check-dvs.py` only scans each block at a time, so that parser is not usable. And it doesn't look like `metamath-knife`'s API has the ability to modify a database file: https://docs.rs/metamath-rs/0.3.8/metamath_rs/database/struct.Database.html

Oftentimes a hypothesis can be removed by an `eqid` (or `cbv`). But this will require more extensive automation work, and the easiest initial automatic solution will probably keep the old theorem as a lemma.
